Max retries

ColinFinck opened the pull request:
As suggested in #184, I've fixed the infinite loop when reading from a broken Modbus RTU bus by introducing a maximum number of retries. When that number is reached, the code gives up and returns an error to the caller.

`MAX_RETRIES` is a constant that is set to 20 for now. Feel free to change that or make it configurable.

I had to edit other parts of the RTU code, which assumed that decoding can never fail. I took the opportunity to make them more readable and more consistent with the TCP code. Should get obvious when you check this PR commit by commit.

Let me know if this can go through or if there's anything I have to change.

Fixes #184

uklotzde commented on src/codec/rtu.rs:
```diff
@@ -36,42 +36,47 @@ impl FrameDecoder {
         buf: &mut BytesMut,
         pdu_len: usize,
     ) -> Result<Option<(SlaveId, Bytes)>> {
+        const CRC_BYTE_COUNT: usize = 2;
```
:+1:

uklotzde commented on src/codec/rtu.rs:
```diff
@@ -36,42 +36,47 @@ impl FrameDecoder {
-                },
-                Err(err) => Err(err),
+
+        if buf.len() < adu_len + CRC_BYTE_COUNT {
```
I am also an advocate of using early returns to keep the main path straight.

uklotzde commented on src/codec/rtu.rs:
```diff
@@ -36,42 +36,47 @@ impl FrameDecoder {
+        let crc_buf = buf.split_to(CRC_BYTE_COUNT);
+
+        // Read trailing CRC and verify ADU
+        match Cursor::new(&crc_buf)
```
Since the `Ok` branch doesn't carry any value we we could use any early return for the error case? Would be consistent with the early return above. On the other hand the current code verifies that the unit type and nothing else is returned on `Ok`. I am fine with both, your choice.

ColinFinck replied:
I changed that to an early return. Saves us one level of indentation :)

uklotzde commented on src/codec/rtu.rs:
```diff
@@ -238,60 +243,51 @@ fn decode<F>(
     }
+
+    // Maximum number of retries exceeded.
+    log::error!("Giving up to decode frame after {} retries", MAX_RETRIES);
```
`{MAX_RETRIES}`

ColinFinck replied:
Changed. I refrained from doing that as I didn't know about your MSRV policy. The only hint was `edition = "2021"` indicating that I can use all features of Rust 1.56.
Capturing identifiers in format strings has increased the Minimum Supported Rust Version to 1.58.

uklotzde commented on src/codec/rtu.rs:
```diff
@@ -300,36 +296,28 @@ impl Decoder for ServerCodec {
-                // never fail!
-                unreachable!();
-            })
+        if let Some((slave_id, pdu_data)) = self.decoder.decode(buf)? {
```
A `let-else` binding with an early return would improve readability.

ColinFinck replied:
Changed. Like above, I refrained from introducing that earlier, because it's only supported since Rust 1.65.
Even worse, while it's supported, it's currently not picked up by `rustfmt`, so people will introduce their own coding styles around let-else. I've written it in a way that I consider meaningful, but a future `rustfmt` may reformat that.

Along with this change, I've added a line `rust-version = "1.65"` to `Cargo.toml` and added an entry to the changelog.

uklotzde reviewed (commented):
Thank you very much! Just some minor comments/questions.

Please rebase on main, I have just fixed the unrelated pre-commit issue.

uklotzde commented:
While you are at it: Please add an entry to the changelog. I guess it is time to finally release this new version.

ColinFinck commented:
I've addressed all your comments.
If you still have that Serial-to-USB adapter, which originally made you skip errors and implement error recovery, please verify that my new code still works with that. Other than that, it should be good to go.
